sources/constraint-declaration-validation.asciidoc

```diff
@@ -1026,6 +1100,289 @@ public class SimpleOrderService extends OrderService {
+----
+private List<@NotEmpty ?> list;
+----
+====
+
+[[constraintdeclarationvalidationprocess-containerelementconstraints-implicitunwrapping]]
+==== Implicit unwrapping of containers
+
+Besides specifying container element constraints on type arguments, it is also possible to declare container element constraints on non-generic container types.
+This is done by means of _implicit unwrapping_, i.e. a constraint doesn't apply to the annotated container itself but to its elements.
+An example are JavaFX's non-generic property types such as `StringProperty` or `IntegerProperty`:
+
+.Implicit unwrapping of container elements
+====
+[source, JAVA]
+----
+@Size(max=10) private StringProperty name;
+@Positive private IntegerProperty positiveInt;
+----
+====
+
+Here the `@Size` and `@Positive` constraints don't apply to the annotated `StringProperty` and `IntegerProperty` themselves, but rather to the wrapped string and integer values, respectively.
+
+For this to work, an unambiguously resolvable value extractor must be defined which carries the `@UnwrapByDefault` annotation (see <<constraintdeclarationvalidationprocess-valueextractordefinitions-unwrapbydefault>>).
+
+If needed, the target (container or container element) of a constraint declared on a container can be explicitly specified via the `Unwrap` and `Skip` payload definitions:
+
+.`Payload` types for unwrapping control
+====
+[source, JAVA, indent=0]
+----
+include::{validation-api-source-dir}javax/validation/valueextraction/Unwrapping.java[lines=7..8;11..-1]
+----
+====
+
+This is useful for applying a constraint given on a non-generic container to
+
+* the container element if there is no value extractor marked with `@UnwrapByDefault` (by using `Unwrap`)
+* the container itself in case there is a value extractor marked with `@UnwrapByDefault` (by using `Skip`)
+
+For instance the `@NotNull` constraint is applied to the `StringProperty` container:
+
+[source,java]
+----
+@NotNull(payload = Unwrapping.Skip.class) StringProperty name;
+----
+
+[NOTE]
+====
+For the sake of readability, when applying constraints to the elements of a generic container,
+it is strongly recommended to put the constraints to the type argument instead of the container itself in conjunction with `Unwrapping.Unwrap`.
+I.e. you should prefer
+
+[source,java]
+----
+List<@Email String> emails;
+----
+
+over
+
+[source,java]
+----
+@Email(payload = Unwrapping.Unwrap.class)
+List<String> emails;
+----
+====
+
+[[constraintdeclarationvalidationprocess-valueextractordefinitions]]
+=== Value extractor definitions
+
+Validation of container element constraints as well as cascaded validation of generic container types requires access to the values stored in the container.
+The retrieval of values stored in a container is handled via implementations of the `ValueExtractor` interface:
+
+[source, JAVA, indent=0]
+----
+include::{validation-api-source-dir}javax/validation/valueextraction/ValueExtractor.java[lines=7..8;15..-1]
```

Comment:
Also we say 
"@param nodeName the name of the node to be added to the {@link Path}"

It should be more explicit
"@param nodeName the name of the node representing the container element. It will be added to the {@link Path}"